### PR TITLE
Document .validate_templates()

### PR DESCRIPTION
**Pull request by dekatzenel**
# Citrine Python PR

## Description 
Document .validate_templates()

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

### Adherence to team decisions
- [ ] I have added tests for 100% coverage
- [ ] I have written Numpy-style docstrings for every method and class.
- [ ] I have communicated the downstream consequences of the PR to others.
- [x] I have bumped the version in setup.py

### CODE REVIEW

**Comment by bfolie** [comment]
Could you please make it one sentence per line (for ease of future PR review)?

File: docs/source/data_entry.rst [hunk 1]
```diff
@@ -129,4 +132,44 @@ and if any fail, provide the same error that the method would provide without th
...
 object will not be registered or deleted.
 
-Setting ``dry_run=False`` is equivalent to not specifying ``dry_run`` at all and will have no effect.
\ No newline at end of file
+Setting ``dry_run=False`` is equivalent to not specifying ``dry_run`` at all and will have no effect.
+
+Template and Simple Validations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Sometimes, it is convenient to be able to bulk validate a group of runs and/or specs against their attribute and object
+templates before any of the data objects is stored. The ``.validate_templates()`` methods, available for all runs and
```
Comment (bfolie) [inline docs/source/data_entry.rst:140]:
```suggestion
templates before any of the data objects are stored. The ``.validate_templates()`` methods, available for all runs and
```

File: docs/source/data_entry.rst [hunk 2]
```diff
@@ -129,4 +132,44 @@ and if any fail, provide the same error that the method would provide without th
...
+    dataset.process_runs.validate_templates(run, object_template=template)
+
+Example usage for an ingredient:
+
+.. code-block:: python
+
+    process_template = ProcessTemplate("pt")
+    process_spec = ProcessSpec("ps", template=process_template)
+    material_spec = MaterialSpec("ms", process=process_spec)
+    ingredient_spec = IngredientSpec("is", process=process_spec, material=material_spec)
```
Comment (bfolie) [inline docs/source/data_entry.rst:174]: Isn't this creating a loop (the process makes the material which is used as an ingredient in the process)?

File: docs/source/data_entry.rst [hunk 3]
```diff
@@ -129,4 +132,44 @@ and if any fail, provide the same error that the method would provide without th
...
+    dataset.ingredient_specs.validate_templates(run, ingredient_process_template=process_template)
```
Comment (bfolie) [inline docs/source/data_entry.rst:175]: What's `run`?

File: docs/source/data_entry.rst [hunk 4]
```diff
@@ -129,4 +132,44 @@ and if any fail, provide the same error that the method would provide without th
...
+
+Template and Simple Validations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Sometimes, it is convenient to be able to bulk validate a group of runs and/or specs against their attribute and object
+templates before any of the data objects is stored. The ``.validate_templates()`` methods, available for all runs and
+specs, validate the provided object against already-stored attribute templates as well as a provided object template.
+Notably, these methods do not validate linked objects in any way, making it possible to run validations on an object
+with links to yet-unstored objects. Be aware that this means that ``.validate_templates()`` does not validate links and
+will not surface any link-based errors. For ingredients, the associated object template is a process template that is
+provided as a separate parameter. This method returns a list of validation errors, which is empty on validation success.
```
Comment (bfolie) [inline docs/source/data_entry.rst:145]: The sentence about ingredients seems like a secondary set of information: once the reader understands validation in the general case, here's an extra wrinkle for the ingredients case. I suggest it be separated. After explaining and showing examples for typical validation, then explain and show an example for ingredient validation.

File: docs/source/data_entry.rst [hunk 5]
```diff
@@ -129,4 +132,44 @@ and if any fail, provide the same error that the method would provide without th
...
 
-Setting ``dry_run=False`` is equivalent to not specifying ``dry_run`` at all and will have no effect.
\ No newline at end of file
+Setting ``dry_run=False`` is equivalent to not specifying ``dry_run`` at all and will have no effect.
+
+Template and Simple Validations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Sometimes, it is convenient to be able to bulk validate a group of runs and/or specs against their attribute and object
+templates before any of the data objects is stored. The ``.validate_templates()`` methods, available for all runs and
+specs, validate the provided object against already-stored attribute templates as well as a provided object template.
```
Comment (bfolie) [inline docs/source/data_entry.rst:141]: Where do these attribute templates come from? Are links to them expected to be in the object template?

**Comment by bfolie** [comment]
Could you expand on the examples? They show the structure of the methods, but their usage would make more sense if some example outputs were shown. Since the successful case results in an empty list, there should be examples of failed validation.

File: docs/source/data_entry.rst (hunk 5 above)
Comment (dekatzenel) [reply docs/source/data_entry.rst]: Not necessarily. Links are expected to be on the object

File: docs/source/data_entry.rst [hunk 6]
```diff
@@ -137,39 +137,75 @@ Setting ``dry_run=False`` is equivalent to not specifying ``dry_run`` at all and
...
     process_spec = ProcessSpec("ps", template=process_template)
-    material_spec = MaterialSpec("ms", process=process_spec)
-    ingredient_spec = IngredientSpec("is", process=process_spec, material=material_spec)
-    dataset.ingredient_specs.validate_templates(run, ingredient_process_template=process_template)
+
+    mat_process_spec = ProcessSpec("mps")
+    material_spec = MaterialSpec("ms", process=mat_process_spec)
+
+    ingredient_spec = IngredientSpec("is", process=process_spec, material=material_spec, labels=["ingredient"])
+    dataset.ingredient_specs.validate_templates(ingredient_spec, ingredient_process_template=process_template)
```
Comment (bfolie) [inline docs/source/data_entry.rst:204]: If the collection is ingredient specs/runs, is the `object_template` field ignored?

Comment (dekatzenel) [reply docs/source/data_entry.rst:204]: Sort of? It's passed through to the core validation code, which then validates the object template against all attributes that have attribute templates that are also on the object template. But ingredients don't have attributes, so there's nothing to validate the object template against

File: docs/source/data_entry.rst [hunk 7]
```diff
@@ -129,4 +132,80 @@ and if any fail, provide the same error that the method would provide without th
 validations succeed, the method returns the same success value that it would without the ``dry_run`` argument, but the
 object will not be registered or deleted.
 
-Setting ``dry_run=False`` is equivalent to not specifying ``dry_run`` at all and will have no effect.
\ No newline at end of file
+Setting ``dry_run=False`` is equivalent to not specifying ``dry_run`` at all and will have no effect.
+
+Template and Simple Validations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Sometimes, it is convenient to be able to bulk validate a group of runs and/or specs against their attribute and object
```
Comment (bfolie) [inline docs/source/data_entry.rst:139]:
```suggestion
Sometimes, it is convenient to validate a group of runs and/or specs against their attribute and object
```

File: docs/source/data_entry.rst [hunk 8]
```diff
@@ -129,4 +132,80 @@ and if any fail, provide the same error that the method would provide without th
...
+templates before any of the data objects are stored.
+The ``.validate_templates()`` methods, available for all runs and specs, validate the provided object against all of the
+(already-stored) attribute templates linked to attributes on the object as well as an optional provided object template.
```
Comment (bfolie) [inline docs/source/data_entry.rst:142]:
```suggestion
(already-stored) attribute templates linked to attributes on the object as well as against an optional object template.
```

File: docs/source/data_entry.rst [hunk 9]
```diff
@@ -129,4 +132,80 @@ and if any fail, provide the same error that the method would provide without th
...
+Notably, these methods do not validate linked objects in any way, making it possible to run validations on an object
+with links to yet-unstored objects.
+Be aware that this means that ``.validate_templates()`` does not validate links and will not surface any link-based
```
Comment (bfolie) [inline docs/source/data_entry.rst:145]:
```suggestion
Be aware that this means that ``.validate_templates()`` will not surface any link-based
```

File: docs/source/data_entry.rst [hunk 10]
```diff
@@ -129,4 +132,80 @@ and if any fail, provide the same error that the method would provide without th
...
+errors.
```
Comment (bfolie) [inline docs/source/data_entry.rst:146]:
```suggestion
Be aware that this means that ``.validate_templates()`` will not surface any link-based errors.
```

File: docs/source/data_entry.rst [hunk 11]
```diff
@@ -129,4 +132,80 @@ and if any fail, provide the same error that the method would provide without th
...
+This method returns a list of validation errors, which is empty on validation success.
+
+The below example usages are trivial examples intended to illustrate the method arguments, rather than interesting
+validation cases.
```
Comment (bfolie) [inline docs/source/data_entry.rst:150]:
```suggestion
The examples below illustrate the usage of ``.validate_templates()``.
```

**Comment by bfolie** [comment]
I made stylistic suggestions. You can incorporate them as you see fit and then I'll approve

**Pull request merged**